File: AndroidWebBox/coconut-plugins/src/main/java/com/sniper/coconut/components/network/NetworkComponent.kt
package com.sniper.coconut.components.network

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import com.sniper.coconut.component.BaseComponent
import com.sniper.coconut.component.ComponentContext
import com.sniper.coconut.component.ComponentMetadata
import com.sniper.coconut.utils.Logger
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

/**
 * Network Component (Enhanced)
 *
 * Provides:
 * - Network status checking
 * - Native HTTP request proxy for H5 (bypasses CORS)
 * - Cookie sync support
 * - Auto-injected headers (token, version, device info)
 *
 * HTTP implementation is auto-selected at runtime:
 * - OkHttp (preferred) when available in host app classpath
 * - HttpURLConnection (fallback) when OkHttp is not present
 *
 * H5 Usage:
 *   Coconut.call('network.request', {
 *     url: 'https://api.example.com/data',
 *     method: 'GET',
 *     headers: { 'X-Custom': 'value' },
 *     body: null,
 *     timeout: 15000
 *   }, callback)
 */
@ComponentMetadata(
    name = "network",
    version = "2.0.0",
    description = "Network status and native HTTP proxy component"
)
class NetworkComponent : BaseComponent() {

    override val name = "network"
    override val version = "2.0.0"
    override val description = "Network status and native HTTP proxy component"

    private var componentContext: ComponentContext? = null
    private val connectivityManager: ConnectivityManager?
        get() = componentContext?.applicationContext?.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager

    // Extra headers to auto-inject (e.g. token, version)
    private val extraHeaders = mutableMapOf<String, String>()

    private var httpClient: HttpClient? = null

    override suspend fun onInit(ctx: ComponentContext) {
        componentContext = ctx
        httpClient = HttpClientFactory.create()
        Logger.d(name, "Using ${httpClient?.javaClass?.simpleName}")
    }

    override suspend fun handle(function: String, params: JsonObject?): JsonElement {
        return when (function) {
            "getType" -> getNetworkType()
            "getState" -> getNetworkState()
            "isConnected" -> isConnected()
            "request" -> httpRequest(params)
            "get" -> httpGet(params)
            "post" -> httpPost(params)
            else -> functionNotSupportedError(function)
        }
    }

    // ---- Network Status ----

    private fun getNetworkType(): JsonElement {
        val capabilities = getNetworkCapabilities()
        val type = when {
            capabilities == null -> "none"
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> "wifi"
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> "cellular"
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_VPN) -> "vpn"
            else -> "unknown"
        }
        return buildJsonObject { put("type", JsonPrimitive(type)) }.let { success(it) }
    }

    private fun getNetworkState(): JsonElement {
        val capabilities = getNetworkCapabilities()
        return buildJsonObject {
            put("isConnected", JsonPrimitive(capabilities != null))
            put("type", JsonPrimitive(
                when {
                    capabilities == null -> "none"
                    capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> "wifi"
                    capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> "cellular"
                    else -> "unknown"
                }
            ))
        }.let { success(it) }
    }

    private fun isConnected(): JsonElement {
        val connected = getNetworkCapabilities() != null
        return buildJsonObject { put("isConnected", JsonPrimitive(connected)) }.let { success(it) }
    }

    private fun getNetworkCapabilities(): NetworkCapabilities? {
        val cm = connectivityManager ?: return null
        val network = cm.activeNetwork ?: return null
        return cm.getNetworkCapabilities(network)
    }

    // ---- Native HTTP Proxy ----

    /**
     * Generic HTTP request (H5 calls this to bypass CORS)
     * Delegates to HttpClient implementation (OkHttp or HttpURLConnection)
     */
    private suspend fun httpRequest(params: JsonObject?): JsonElement {
        val url = getParam(params, "url")
        if (url.isEmpty()) return error("900001", "url is required")

        val client = httpClient ?: HttpClientFactory.create()

        val method = getParam(params, "method", "GET").uppercase()
        val timeout = getIntParam(params, "timeout", 15000)
        val contentType = getParam(params, "contentType", "application/json")
        val body = getParam(params, "body", "")

        // Merge extra headers + H5-provided headers
        val headers = mutableMapOf<String, String>()
        headers.putAll(extraHeaders)
        params?.get("headers")?.let { headersElem ->
            if (headersElem is JsonObject) {
                headersElem.forEach { (k, v) ->
                    if (v is JsonPrimitive && v.isString) {
                        headers[k] = v.content
                    }
                }
            }
        }

        Logger.d(name, "Native request: $method $url")

        val request = HttpRequest(
            url = url,
            method = method,
            headers = headers,
            body = body.ifEmpty { null },
            contentType = contentType,
            timeout = timeout
        )

        val response = client.execute(request)

        return if (response.statusCode == -1) {
            buildJsonObject {
                put("statusCode", JsonPrimitive(-1))
                put("error", JsonPrimitive(response.body))
            }.let { success(it) }
        } else {
            buildJsonObject {
                put("statusCode", JsonPrimitive(response.statusCode))
                put("body", JsonPrimitive(response.body))
                put("headers", JsonPrimitive(response.headers.toString()))
            }.let { success(it) }
        }
    }

    /**
     * Shorthand GET
     */
    private suspend fun httpGet(params: JsonObject?): JsonElement {
        val mergedParams = buildJsonObject {
            params?.forEach { (k, v) -> put(k, v) }
            put("method", JsonPrimitive("GET"))
        }
        return httpRequest(mergedParams)
    }

    /**
     * Shorthand POST
     */
    private suspend fun httpPost(params: JsonObject?): JsonElement {
        val mergedParams = buildJsonObject {
            params?.forEach { (k, v) -> put(k, v) }
            put("method", JsonPrimitive("POST"))
        }
        return httpRequest(mergedParams)
    }

    /**
     * Add extra header to auto-inject
     */
    fun addExtraHeader(key: String, value: String) {
        extraHeaders[key] = value
    }

    override suspend fun onCleanup() {
        componentContext = null
        extraHeaders.clear()
        httpClient = null
    }
}

File: iOSWebBox/CoconutSDK/Sources/CoconutSDK/Components/NetworkComponent.swift
import Foundation
import Network

public class NetworkComponent: BaseComponent {

    override public var name: String { "network" }
    override public var version: String { "2.0.0" }
    override public var pluginDescription: String { "Network status and native HTTP proxy component" }

    private var extraHeaders: [String: String] = [:]

    override public func handle(function: String, params: [String: Any]?) async throws -> [String: Any] {
        switch function {
        case "getType": return getNetworkType()
        case "getState": return getNetworkState()
        case "isConnected": return isConnected()
        case "request": return try await httpRequest(params)
        case "get": return try await httpGet(params)
        case "post": return try await httpPost(params)
        default: try functionNotSupportedError(function)
        }
    }

    private func getNetworkType() -> [String: Any] {
        let type: String
        if let connection = NetworkMonitor.shared.currentType {
            switch connection {
            case .wifi: type = "wifi"
            case .cellular: type = "cellular"
            case .vpn: type = "vpn"
            }
        } else {
            type = "none"
        }
        return success(["type": type])
    }

    private func getNetworkState() -> [String: Any] {
        let connected = NetworkMonitor.shared.isConnected
        let type: String
        if let connection = NetworkMonitor.shared.currentType {
            switch connection {
            case .wifi: type = "wifi"
            case .cellular: type = "cellular"
            case .vpn: type = "vpn"
            }
        } else {
            type = connected ? "unknown" : "none"
        }
        return success(["isConnected": connected, "type": type])
    }

    private func isConnected() -> [String: Any] {
        return success(["isConnected": NetworkMonitor.shared.isConnected])
    }

    private func httpRequest(_ params: [String: Any]?) async throws -> [String: Any] {
        let url = getParam(params, "url")
        if url.isEmpty { try error("200007", "url is required") }

        let method = getParam(params, "method", "GET").uppercased()
        let timeout = getIntParam(params, "timeout", 15000)
        let contentType = getParam(params, "contentType", "application/json")
        let body = getParam(params, "body", "")

        var headers = extraHeaders
        if let h5Headers = params?["headers"] as? [String: Any] {
            for (k, v) in h5Headers {
                if let str = v as? String { headers[k] = str }
            }
        }

        Logger.shared.d(name, "Native request: \(method) \(url)")

        do {
            guard let requestUrl = URL(string: url) else {
                try error("200007", "Invalid URL: \(url)")
            }

            var request = URLRequest(url: requestUrl)
            request.httpMethod = method
            request.timeoutInterval = Double(timeout) / 1000.0

            for (k, v) in headers {
                request.setValue(v, forHTTPHeaderField: k)
            }

            if method != "GET" && !body.isEmpty {
                request.httpBody = body.data(using: .utf8)
                request.setValue(contentType, forHTTPHeaderField: "Content-Type")
            }

            let (data, response) = try await URLSession.shared.data(for: request)

            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? -1
            let responseBody = String(data: data, encoding: .utf8) ?? ""

            var responseHeaders: [String: String] = [:]
            if let httpRes = response as? HTTPURLResponse {
                for (key, value) in httpRes.allHeaderFields {
                    if let keyStr = key as? String, let valueStr = value as? String {
                        responseHeaders[keyStr] = valueStr
                    }
                }
            }

            Logger.shared.d(name, "Response: \(statusCode) (\(responseBody.count) bytes)")

            return success([
                "statusCode": statusCode,
                "body": responseBody,
                "headers": "\(responseHeaders)"
            ])
        } catch {
            Logger.shared.e(name, "Request failed: \(url)", error)
            return success([
                "statusCode": -1,
                "error": error.localizedDescription
            ])
        }
    }

    private func httpGet(_ params: [String: Any]?) async throws -> [String: Any] {
        var merged = params ?? [:]
        merged["method"] = "GET"
        return try await httpRequest(merged)
    }

    private func httpPost(_ params: [String: Any]?) async throws -> [String: Any] {
        var merged = params ?? [:]
        merged["method"] = "POST"
        return try await httpRequest(merged)
    }

    public func addExtraHeader(_ key: String, _ value: String) {
        extraHeaders[key] = value
    }

    override public func onCleanup() async {
        extraHeaders.removeAll()
    }
}

class NetworkMonitor {
    static let shared = NetworkMonitor()

    enum ConnectionType {
        case wifi, cellular, vpn
    }

    private(set) var isConnected = false
    private(set) var currentType: ConnectionType?

    private let monitor = NWPathMonitor()
    private let queue = DispatchQueue(label: "com.coconut.networkmonitor")

    private init() {
        monitor.pathUpdateHandler = { [weak self] path in
            self?.isConnected = path.status == .satisfied
            if path.usesInterfaceType(.wifi) {
                self?.currentType = .wifi
            } else if path.usesInterfaceType(.cellular) {
                self?.currentType = .cellular
            } else if path.usesInterfaceType(.other) {
                self?.currentType = .vpn
            } else {
                self?.currentType = nil
            }
        }
        monitor.start(queue: queue)
    }
}
